File: ABC/abc049/c/main.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

string words[] = {"dream", "dreamer", "erase", "eraser"};
int nWords = sizeof(words) / sizeof(words[0]);

int main() {
    string S; cin >> S;
    reverse(S.begin(), S.end());
    for (int i = 0; i < nWords; ++i) {
        reverse(words[i].begin(), words[i].end());
    }

    for (int i = 0, max = S.size(); i < max; ) {
        bool found = false;
        for (int j = 0; j < nWords; ++j) {
            string word = words[j];
            if (S.substr(i, word.size()) == word) {
                i += word.size();
                found = true;
                break;
            }
        }
        if (!found) {
            cout << "NO" << endl;
            return 0;
        }
    }
    cout << "YES" << endl;
}

File: ABC/abc049/c/main.py
#!/usr/bin/env python

def solve():
    s = input()
    rev_s = s[::-1]
    words = ['dream', 'dreamer', 'erase', 'eraser']
    rev_words = [i[::-1] for i in words]

    i = 0
    max = len(s)
    while i < max:
        found = False
        for word in rev_words:
            if rev_s[i:].startswith(word):
                found = True
                i += len(word)
                break
        if not found:
            return "NO"

    return "YES"

print(solve())
